torch/distributed/distributed_c10d.py

```diff
@@ -233,6 +233,17 @@ def register_backend(cls, name, func, extended_api=False):
 dist_backend = Backend
 
 
+# NOTE(crcrpar): [ReduceOp static class attributes to support `isinstance`]
+#   A ReduceOp instance of `PREMUL_SUM` is supposed to be created via `_make_nccl_premul_sum`
+#   while the other `op`s (meaning RedOpType members) can be directly passed to c10d reduce collectives.
+#   I changed `ReduceOp` to struct from enum class and introduced RedOpType enum class for PREMUL_SUM,
+#   which broke an implicit contract of ReduceOp being enum-like with which users apply isinstance to
+#   `op`, for example, `isinstance(ReduceOp.SUM, ReduceOp)`: https://github.com/pytorch/pytorch/issues/87191
+DENY_LIST = ("PREMUL_SUM", )
+for _red_op_name, _red_op_value in ReduceOp.RedOpType.__members__.items():
```

Comment:
@crcrpar I think this PR break the serialization of c10d.ReduceOp, with the newest nightly this now crashes:
```
import torch
import copy

a = torch.distributed.distributed_c10d.ReduceOp.SUM
copy.deepcopy(a)
```
I checked the nightly a few days ago and it works all fine, and the issue seems coming from the changes in this file, if I comment this changes, then it works again, could you help take a look on this and fix it? Thanks!

Reply:
I think this might accidentally copies the `ReduceOp.RedOpType` get_state set_state or sth to the `ReduceOp`, we should probably only copy the ENUM values

Reply:
I'm wondering if we should try to override `__isinstance__` instead of overriding the attributes on class?

Reply:
I can confirm that I'm also running into the issue @wanchaol mentioned